### PR TITLE
Skippable facts

### LINKED ISSUE #263: UpdateLocalPositionAndRotationMethod unit test should check unity version
#### Bug description
SetLocalPositionAndRotationTests.UpdateLocalPositionAndRotationMethod Fails if unity version is too old

- Version of analyzers assembly: 79a79a9
- Analyzer rule: SetLocalPositionAndRotationTests.UpdateLocalPositionAndRotationMethod
- Error (exception message, type, and callstack where applicable): 
```
[xUnit.net 00:00:03.83]     Microsoft.Unity.Analyzers.Tests.SetLocalPositionAndRotationTests.UpdateLocalPositionAndRotationMethod [FAIL]
  Failed Microsoft.Unity.Analyzers.Tests.SetLocalPositionAndRotationTests.UpdateLocalPositionAndRotationMethod [802 ms]
  Error Message:
   Mismatch between number of diagnostics returned, expected "1" actual "0"

Diagnostics:
    NONE.

Expected: True
Actual:   False
  Stack Trace:
     at Microsoft.Unity.Analyzers.Tests.DiagnosticVerifier.VerifyDiagnosticResults(Diagnostic[] actualResults, DiagnosticAnalyzer analyzer, DiagnosticResult[] expectedResults) in /home/florian/src-misc/roslyn-analyzers/Microsoft.Unity.Analyzers/src/Microsoft.Unity.Analyzers.Tests/Infrastructure/DiagnosticVerifier.cs:line 94
   at Microsoft.Unity.Analyzers.Tests.DiagnosticVerifier.VerifyDiagnosticsAsync(AnalyzerVerificationContext context, String[] sources, DiagnosticAnalyzer analyzer, DiagnosticResult[] expected) in /home/florian/src-misc/roslyn-analyzers/Microsoft.Unity.Analyzers/src/Microsoft.Unity.Analyzers.Tests/Infrastructure/DiagnosticVerifier.cs:line 82
   at Microsoft.Unity.Analyzers.Tests.SetLocalPositionAndRotationTests.UpdateLocalPositionAndRotationMethod() in /home/florian/src-misc/roslyn-analyzers/Microsoft.Unity.Analyzers/src/Microsoft.Unity.Analyzers.Tests/SetLocalPositionAndRotationTests.cs:line 32
--- End of stack trace from previous location ---

Failed!  - Failed:     1, Passed:   291, Skipped:     0, Total:   292, Duration: 7 s - Microsoft.Unity.Analyzers.Tests.dll (net7.0)


```
#### To Reproduce
Steps or code to reproduce the behavior:

1. only have older version of unity installed e.g. 2021.3.5f1

#### Expected behavior
the test should not run

-or-

the test should expect no diagnostic as the method it suggests does not exist

#### Additional context
linux, self built

### PR DESCRIPTION
**Pull request by sailro**
Fixes #263 

#### Checklist

- [X] I have read the [Contribution Guide](/CONTRIBUTING.md) ;
- [X] I have added tests that prove my fix is effective or that my feature works ;
- [X] I have added necessary documentation (if appropriate) ;

#### Short description of what this resolves:
Add skippable facts support to conditionally ignore tests.

Example with 2021.3.9f1 (`Transform.SetLocalPositionAndRotation` was added in 2021.3.11f1)

This is only useful when full analyzer compilation is not needed. For other advanced usages we have `AnalyzerVerificationContext` to filter given options, ids or LanguageVersion (that we use for suppressor tests). 

image

### CODE REVIEW

File: src/Microsoft.Unity.Analyzers.Tests/BaseTest.cs [hunk 1]
```diff
@@ -70,13 +74,32 @@ public async Task DoNotFailWithInterfaceMembers()
...
 
-	protected override CodeFixProvider GetCSharpCodeFixProvider()
+	protected override TCodeFix GetCSharpCodeFixProvider()
 	{
 		return new TCodeFix();
 	}
+
+	protected bool MethodExists(string typeName, string methodName, string assenblyNameFilter = "")
+	{
+		foreach (var assemblyFile in UnityAssemblies().Where(n => n.Contains(assenblyNameFilter)))
```
Comment (jbevain) [inline src/Microsoft.Unity.Analyzers.Tests/BaseTest.cs:89]: the Where + Contains feels unnecessary right now as we're only using it for one case where we're looking for 1 assembly in particular. I'd just Select the one assembly we're looking for and look into that.

File: src/Microsoft.Unity.Analyzers.Tests/BaseTest.cs [hunk 2]
```diff
@@ -70,13 +74,32 @@ public async Task DoNotFailWithInterfaceMembers()
...
+		{
+			var assembly = Assembly.LoadFile(assemblyFile);
```
Comment (jbevain) [inline src/Microsoft.Unity.Analyzers.Tests/BaseTest.cs:91]: Doing a full `Assembly.LoadFile` on Unity assembly seems fragile and we're never going to execute code into that. Maybe use a reflection only assembly load context or Cecil?

File: src/Microsoft.Unity.Analyzers.Tests/SetLocalPositionAndRotationTests.cs [hunk 3]
```diff
@@ -27,6 +27,11 @@ void Update()
 }
 ";
 
+		var method = GetCSharpDiagnosticAnalyzer().ExpressionContext.SetPositionAndRotationMethodName;
+		var type = typeof(UnityEngine.Transform);
+
+		Skip.IfNot(MethodExists(type.FullName!, method, type.Namespace!), $"This Unity version does not support {type}.{method}");
```
Comment (jbevain) [inline src/Microsoft.Unity.Analyzers.Tests/SetLocalPositionAndRotationTests.cs:33]: using `type.Namespace` to get the assembly name for filtering looks weird, let's be explicit and say something like `type.Assembly.GetName().Name`.

Comment (sailro) [reply src/Microsoft.Unity.Analyzers.Tests/SetLocalPositionAndRotationTests.cs:33]: `type.Assembly.GetName().Name` is the analyzer assembly given we use stubs, going to use "UnityEngine" :)

File: src/Microsoft.Unity.Analyzers.Tests/BaseTest.cs (hunk 2 above)
Comment (sailro) [reply src/Microsoft.Unity.Analyzers.Tests/BaseTest.cs:91]: Going to use `AssemblyLoadContext`, `Assembly.ReflectionOnlyLoad` is not a thing anymore in .NET Core.

File: src/Microsoft.Unity.Analyzers.Tests/BaseTest.cs [hunk 4]
```diff
@@ -70,13 +73,36 @@ public async Task DoNotFailWithInterfaceMembers()
...
-	protected override CodeFixProvider GetCSharpCodeFixProvider()
+	protected override TCodeFix GetCSharpCodeFixProvider()
 	{
 		return new TCodeFix();
 	}
+
+	protected bool MethodExists(string assemblyName, string typeName, string methodName)
+	{
+		var assemblyFilePath = UnityAssemblies().First(f => Path.GetFileNameWithoutExtension(f) == assemblyName);
+		var ctx = new AssemblyLoadContext(assemblyName, isCollectible: true);
```
Comment (jbevain) [inline src/Microsoft.Unity.Analyzers.Tests/BaseTest.cs:89]: `MetadataLoadContext` instead of AssemblyLoadContext.

**Review by jbevain — APPROVED** [review]

**Pull request merged**